File: src/main/java/fr/adrienbrault/idea/symfonyplugin/intentions/xml/XmlServiceSuggestIntention.java
package fr.adrienbrault.idea.symfonyplugin.intentions.xml;

import com.intellij.codeInsight.hint.HintManager;
import com.intellij.codeInsight.intention.PsiElementBaseIntentionAction;
import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.project.Project;
import com.intellij.psi.PsiElement;
import com.intellij.psi.util.PsiTreeUtil;
import com.intellij.psi.xml.XmlTag;
import com.intellij.util.IncorrectOperationException;
import com.intellij.util.containers.ContainerUtil;
import fr.adrienbrault.idea.symfonyplugin.Symfony2ProjectComponent;
import fr.adrienbrault.idea.symfonyplugin.config.xml.XmlHelper;
import fr.adrienbrault.idea.symfonyplugin.intentions.php.XmlServiceArgumentIntention;
import fr.adrienbrault.idea.symfonyplugin.intentions.ui.ServiceSuggestDialog;
import fr.adrienbrault.idea.symfonyplugin.util.dict.ServiceUtil;
import org.apache.commons.lang.StringUtils;
import org.jetbrains.annotations.NotNull;

import java.util.Set;

/**
 * @author Daniel Espendiller <daniel@espendiller.net>
 */
public class XmlServiceSuggestIntention extends PsiElementBaseIntentionAction {

    @Override
    public boolean isAvailable(@NotNull Project project, Editor editor, @NotNull PsiElement psiElement) {
        if(!Symfony2ProjectComponent.isEnabled(psiElement.getProject())) {
            return false;
        }

        final XmlTag argumentTag = PsiTreeUtil.getParentOfType(psiElement, XmlTag.class);
        if(argumentTag == null || !"argument".equals(argumentTag.getName())) {
            return false;
        }

        return XmlServiceArgumentIntention.getServiceTagValid(psiElement) != null;
    }

    @Override
    public void invoke(@NotNull Project project, Editor editor, @NotNull PsiElement psiElement) throws IncorrectOperationException {
        XmlTag xmlTag = XmlServiceArgumentIntention.getServiceTagValid(psiElement);
        if (xmlTag == null) {
            return;
        }

        String aClass = XmlHelper.getClassFromServiceDefinition(xmlTag);
        if(aClass == null || StringUtils.isBlank(aClass)) {
            return;
        }

        final XmlTag argumentTag = PsiTreeUtil.getParentOfType(psiElement, XmlTag.class);
        if(argumentTag == null) {
            return;
        }

        int argumentIndex = XmlHelper.getArgumentIndex(argumentTag);

        Set<String> suggestions = ServiceUtil.getServiceSuggestionsForServiceConstructorIndex(project, aClass, argumentIndex);
        if(suggestions.size() == 0) {
            HintManager.getInstance().showErrorHint(editor, "No suggestion found");
            return;
        }

        ServiceSuggestDialog.create(editor, suggestions, new MyInsertCallback(argumentTag));
    }

    @NotNull
    @Override
    public String getFamilyName() {
        return "Symfony";
    }

    @NotNull
    @Override
    public String getText() {
        return "Symfony: Suggest Service";
    }

    public static class MyInsertCallback implements ServiceSuggestDialog.Callback {
        private final XmlTag argumentTag;

        public MyInsertCallback(XmlTag argumentTag) {
            this.argumentTag = argumentTag;
        }

        @Override
        public void insert(@NotNull String selected) {

            // set type="service" for lazy devs
            if(ContainerUtil.find(argumentTag.getAttributes(), xmlAttribute -> "type".equals(xmlAttribute.getName())) == null) {
                argumentTag.setAttribute("type", "service");
            }

            // append type="SERVICE"
            argumentTag.setAttribute("id", selected);
        }
    }
}

File: src/main/kotlin/fr/adrienbrault/idea/symfony2plugin/intentions/xml/XmlServiceSuggestIntention.kt
package fr.adrienbrault.idea.symfony2plugin.intentions.xml

import com.intellij.codeInsight.intention.PsiElementBaseIntentionAction
import com.intellij.openapi.editor.Editor
import com.intellij.openapi.project.Project
import com.intellij.psi.PsiElement
import com.intellij.psi.util.PsiTreeUtil
import com.intellij.psi.xml.XmlTag
import com.intellij.util.containers.ContainerUtil
import fr.adrienbrault.idea.symfony2plugin.Symfony2ProjectComponent
import fr.adrienbrault.idea.symfony2plugin.action.ServiceActionUtil
import fr.adrienbrault.idea.symfony2plugin.config.xml.XmlHelper
import fr.adrienbrault.idea.symfony2plugin.intentions.ui.ServiceSuggestDialog
import fr.adrienbrault.idea.symfony2plugin.util.IdeHelper
import fr.adrienbrault.idea.symfony2plugin.util.dict.ServiceUtil
import org.apache.commons.lang3.StringUtils

/**
 * @author Daniel Espendiller <daniel@espendiller.net>
 */
open class XmlServiceSuggestIntention : PsiElementBaseIntentionAction() {
    override fun isAvailable(project: Project, editor: Editor?, psiElement: PsiElement): Boolean {
        if (!Symfony2ProjectComponent.isEnabled(project)) {
            return false
        }

        val argumentTag = PsiTreeUtil.getParentOfType(psiElement, XmlTag::class.java)
        if (argumentTag == null || "argument" != argumentTag.name) {
            return false
        }

        return ServiceActionUtil.getServiceTagValid(psiElement) != null
    }

    override fun invoke(project: Project, editor: Editor?, psiElement: PsiElement) {
        val xmlTag = ServiceActionUtil.getServiceTagValid(psiElement)
            ?: return

        val aClass = XmlHelper.getClassFromServiceDefinition(xmlTag)
        if (aClass == null || StringUtils.isBlank(aClass)) {
            return
        }

        val argumentTag = PsiTreeUtil.getParentOfType(psiElement, XmlTag::class.java)
            ?: return

        val argumentIndex = XmlHelper.getArgumentIndex(argumentTag)

        val suggestions = ServiceUtil.getServiceSuggestionsForServiceConstructorIndex(project, aClass, argumentIndex)
        if (suggestions.isEmpty()) {
            IdeHelper.showErrorHintIfAvailable(editor!!, "No suggestion found")
            return
        }

        ServiceSuggestDialog.create(editor!!, suggestions, MyInsertCallback(argumentTag))
    }

    override fun getFamilyName(): String {
        return "Symfony: Suggest Service"
    }

    override fun getText(): String {
        return familyName
    }

    open class MyInsertCallback(private val argumentTag: XmlTag) : ServiceSuggestDialog.Callback {
        override fun insert(selected: String) {
            // set type="service" for lazy devs
            if (ContainerUtil.find(argumentTag.attributes) { xmlAttribute -> "type" == xmlAttribute.name } == null) {
                argumentTag.setAttribute("type", "service")
            }

            // append type="SERVICE"
            argumentTag.setAttribute("id", selected)
        }
    }
}
